compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala

```diff
@@ -194,13 +198,22 @@ class Devalify extends Optimisation {
       } else if (t.symbol.is(CaseAccessor) && !t.symbol.is(Mutable))
         readingOnlyVals(rec)
       else false
+
     case Select(qual, _) if !t.symbol.is(Mutable) =>
-      readingOnlyVals(qual)
+      if (t.symbol == defn.SystemModule) {
+        // System.in is static final fields that, for legacy reasons, must be
+        // allowed to be changed by the methods System.setIn...
+        // https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.4
+        false
+      } else
+        readingOnlyVals(qual)
+
     case t: Ident if !t.symbol.is(Mutable) && !t.symbol.is(Method) && !t.symbol.info.dealias.isInstanceOf[ExprType] =>
```

Comment:
The implementation of this case calls `desugarIdent` and recuse, I think the case where the `Ident` is `System.in` is covered.

@DarkDimius Could you explain why we even need this `desugarIdent`, why not use `Select` everywhere directly? Also where would be the proper was to move that function to factor out the version that's currently copy pasted in the backend?